File: Code-CSharp/Base/Containers/UnorderedSet/UnorderedSetEntry.cs
using System.Collections.Immutable;

namespace CodeCsharp.Base.Containers.UnorderedSet;

public class UnorderedSetEntry {
  public static void UnorderedSetProgram() {
    //initialize a hash set
    HashSet<string> hashSet = new HashSet<string>();
    //initialize a hash set with value
    HashSet<string> hashSet1 = new HashSet<string>{ "aa", "bb", "cc", "dd" };
    //copy hash set
    hashSet = new HashSet<string>(hashSet1);
    //whether hash set is empty
    bool isEmpty = hashSet.Count==0;
    //clear hash set
    hashSet.Clear();
    //insert element
    hashSet.Add("ee");
    //initialize a vector contains values
    List<string> vec = new List<string> { "aa", "bb", "cc", "dd" };
    //insert a range of elements
    hashSet.UnionWith(vec);
    //get size of hash set
    int size = hashSet.Count;
    //whether hash set contains element
    bool b1 = hashSet.Contains("bb");
    //create immutable reference of hash set
    ImmutableHashSet<string> constSetRef = hashSet.ToImmutableHashSet();
    //whether two hash sets equal
    bool b2 = hashSet.OrderBy(k=>k).SequenceEqual(constSetRef.OrderBy(k=>k));
    //remove element equal value
    hashSet.Remove("cc");
  }
  
  public static bool TestAll() {
    bool ret = true;
    HashSet<int> c, c1, c2, c3, c4;
    List<int> vec = new List<int> { 1, 3, 5, 7, 9 };
    c = new HashSet<int>(new List<int> { 1, 2, 3, 4, 5 });
    c2 = new HashSet<int>(c);
    c.Clear();
    c1 = new HashSet<int>(c);
    c.UnionWith(vec);
    c3 = new HashSet<int>(c);
    c.Remove(3);
    c4 = new HashSet<int>(c);
    ret = c1.Count == 0
          && c.Contains(7)
          && c4.Count == 4
          && c2.OrderBy(k=>k).SequenceEqual(
              new HashSet<int>(new List<int> { 1, 2, 3, 4, 5 }.OrderBy(k=>k)))
          && c3.OrderBy(k=>k).SequenceEqual(
              new HashSet<int>(new List<int> { 1, 3, 5, 7, 9 }.OrderBy(k=>k)))
          && c4.OrderBy(k=>k).SequenceEqual(
              new HashSet<int>(new List<int> { 1, 5, 7, 9 }.OrderBy(k=>k)));
    if (!ret)
      Console.WriteLine("Unordered Set Test Failed!");
    return ret;
  }
}

File: Code-Cpp/code_cpp/base/containers/unordered_set/unordered_set_entry.cc
#include "unordered_set_entry.h"
#include <string>
#include <vector>

void UnorderedSetEntry::UnorderedSetProgram() {
  //initialize a hash set
  std::unordered_set<std::string> hash_set;
  //initialize a hash set with value
  std::unordered_set<std::string> hash_set1{"aa", "bb", "cc", "dd"};
  //copy hash set
  hash_set = hash_set1;
  //whether hash set is empty
  bool is_empty = hash_set.empty();
  //clear hash set
  hash_set.clear();
  //insert element
  hash_set.emplace("ee");
  //initialize a vector contains values
  std::vector<std::string> vec{"aa", "bb", "cc", "dd"};
  //insert a range of elements
  hash_set.insert(vec.begin(), vec.end());
  //get size of hash set
  int size = hash_set.size();
  //whether hash set contains element
  bool b1 = hash_set.count("bb") > 0; // set.contains("bb");
  //create immutable reference of hash set
  const std::unordered_set<std::string> &kConstHashSetRef = hash_set;
  //whether two hash sets equal
  bool b2 = hash_set == hash_set1;
  //remove element equal value
  hash_set.erase("dd");
}
bool UnorderedSetEntry::TestAll() {
  bool ret = true;
  using namespace std;
  unordered_set<int> c, c1, c2, c3, c4;
  vector<int> vec{1, 3, 5, 7, 9};
  c = {1, 2, 3, 4, 5};
  c2 = c;
  c.clear();
  c1 = c;
  c.insert(vec.begin(), vec.end());
  c3 = c;
  c.erase(3);
  c4 = c;
  ret = c1.empty()
      && c.count(7) > 0
      && c4.size() == 4
      && c2 == unordered_set<int>{1, 2, 3, 4, 5}
      && c3 == unordered_set<int>{1, 3, 5, 7, 9}
      && c4 == unordered_set<int>{1, 5, 7, 9};
  if (!ret)
    cout << "Hash Set Test Failed!";
  return ret;
}
